File: apps/web/test/clock.test.ts
/**
 * ⭐ İSTEMCİ SAATİ — `serverNow` / `gameNow` / `remaining*` ailesi.
 *
 * **Neden şimdi yazıldı:** bu ailenin hiç testi yoktu ve tam da bu yüzden aynı hata sınıfı
 * canlıda **iki kez** çıktı: 2026-08-02'de casusluk geri sayımı sürekli «varıyor» yazıyordu,
 * 2026-08-07'de asker üretim sayacı kalıcı olarak «sipariş tamamlandı» gösteriyordu. İkisinin
 * de sebebi aynı: **oyun saatindeki bir damgayı gerçek saatle karşılaştırmak.**
 *
 * ⚠️ Asıl işlevi ileriye dönük: Faz 2'de `gameOffsetMs` kaldırılıp `gameNow()` yeniden
 * tanımlanacak. Bu dosya o değişikliğin **sözleşmeyi bozmadığını** ölçen ağdır — özellikle
 * «bitmiş» (`null`) ile «henüz bilmiyorum» ayrımını.
 */
import { beforeEach, describe, expect, it, vi } from 'vitest';
import {
  formatClock, formatDuration, formatLongDuration, gameNow, noteMaintenance,
  noteServerTime, remaining, remainingClock, remainingLong, serverNow, timeAgo,
} from '../src/lib/hooks.ts';

/** Sabit bir "tarayıcı şimdisi" — sapma ölçümü ancak böyle deterministik olur. */
const BROWSER_NOW = Date.parse('2026-08-07T12:00:00.000Z');
const iso = (ms: number): string => new Date(ms).toISOString();

beforeEach(() => {
  vi.useFakeTimers();
  vi.setSystemTime(BROWSER_NOW);
  // Çıpaları her testte bilinen bir noktaya çek (modül düzeyinde durum tutuluyor).
  noteServerTime(iso(BROWSER_NOW), iso(BROWSER_NOW));
});

describe('saat çıpaları', () => {
  it('tarayıcı saati ileriyse sapma ölçülüp geri alınır', () => {
    // Sunucu 30 sn GERİDE → tarayıcı 30 sn ileri demektir.
    noteServerTime(iso(BROWSER_NOW - 30_000));
    expect(serverNow()).toBe(BROWSER_NOW - 30_000);
  });

  it('tarayıcı saati geriyse de düzeltilir', () => {
    noteServerTime(iso(BROWSER_NOW + 45_000));
    expect(serverNow()).toBe(BROWSER_NOW + 45_000);
  });

  /**
   * ⭐⭐ TEK ZAMAN ÇİZGİSİ (0043) — dünya çalışıyorken oyun saati ile gerçek saat AYNI.
   *
   * ⚠️ Eskiden aralarında kalıcı bir fark vardı (dünyanın toplam duraklama süresi, canlıda
   * 196,5 sn) ve bu fark iki kez canlı hataya yol açtı. Artık fark yalnız bakımda oluşuyor.
   */
  it('dünya çalışıyorken gameNow == serverNow', () => {
    noteServerTime(iso(BROWSER_NOW), iso(BROWSER_NOW));
    expect(gameNow()).toBe(serverNow());
  });

  /** Bakımda sunucunun `gameNow`u donmuş `paused_at`i taşır → istemci de donar. */
  it('bakımda gameNow DONAR (sunucunun bildirdiği ana çakılır)', () => {
    const pausedAt = BROWSER_NOW - 10 * 60_000;
    noteServerTime(iso(BROWSER_NOW), iso(pausedAt));
    expect(gameNow()).toBe(pausedAt);

    // Gerçek zaman aksa bile oyun saati kıpırdamaz.
    vi.setSystemTime(BROWSER_NOW + 30_000);
    expect(serverNow()).toBe(BROWSER_NOW + 30_000);
    expect(gameNow()).toBe(pausedAt);
  });

  /** ⚠️ Ağ gecikmesi bakım sanılmamalı — eşiğin altındaki fark gürültüdür. */
  it('küçük fark bakım SAYILMAZ (eşik altı gürültü)', () => {
    noteServerTime(iso(BROWSER_NOW), iso(BROWSER_NOW - 300));
    expect(gameNow()).toBe(serverNow());
  });

  /**
   * ⭐ `noteMaintenance` — WS olayı geldiği anda donar, sunucunun bir sonraki yanıtı
   * beklenmez. Bakım bitince de çözer.
   */
  it('noteMaintenance anında dondurur ve çözer', () => {
    const pausedAt = BROWSER_NOW - 5_000;
    noteMaintenance(true, iso(pausedAt));
    expect(gameNow()).toBe(pausedAt);

    noteMaintenance(false, null);
    expect(gameNow()).toBe(serverNow());
  });

  it('damgasız bakım bildirimi yine de dondurur (akmaya devam etmekten iyi)', () => {
    noteMaintenance(true, null);
    const donmus = gameNow();
    vi.setSystemTime(BROWSER_NOW + 10_000);
    expect(gameNow()).toBe(donmus);
  });

  /** `gameNow` göndermeyen uçlar bakım durumunu BOZMAMALI, yalnız güncellememeli. */
  it('yalnız serverNow gönderen yanıt bakım çıpasını bozmaz', () => {
    const pausedAt = BROWSER_NOW - 60_000;
    noteServerTime(iso(BROWSER_NOW), iso(pausedAt));
    noteServerTime(iso(BROWSER_NOW));               // gameNow yok
    expect(gameNow()).toBe(pausedAt);
  });

  it('bozuk damga çıpayı bozmaz', () => {
    noteServerTime(iso(BROWSER_NOW), iso(BROWSER_NOW));
    noteServerTime('bu bir tarih değil', 'bu da değil');
    expect(gameNow()).toBe(serverNow());
  });
});

describe('remaining — varsayılan çıpa OYUN saati', () => {
  /**
   * ⭐⭐ 2026-08-02 HATASININ BEKÇİSİ — artık YAPISAL olarak imkânsız.
   *
   * Casusluk 120 sn sürüyor. Eski modelde oyun saati gerçek saatin 196,5 sn gerisindeydi, yani
   * varış anı gerçek saatle bakıldığında HEP geçmişte kalıyor ve geri sayım yerine sürekli
   * «varıyor» yazıyordu. Hata görev SÜRESİ kısaldıkça görünür oluyordu — uzun seferlerde
   * yutuluyordu, bu yüzden aylarca fark edilmedi.
   *
   * ⭐ Tek zaman çizgisinde iki çıpa AYNI değeri veriyor: yanlış çıpa seçmek artık bir hataya
   * yol açamıyor. Test bunu kanıtlıyor — ve `gameNow` ile `serverNow` bir gün yeniden ayrışırsa
   * (yalnız bakımda ayrışmalı) burada görünür.
   */
  it('kısa görev (120 sn) doğru sayar — iki çıpa da aynı sonucu verir', () => {
    noteServerTime(iso(BROWSER_NOW), iso(BROWSER_NOW));
    const varis = iso(gameNow() + 120_000);

    expect(remaining(varis)).toBe('2 dk 00 sn');
    expect(remaining(varis, serverNow())).toBe('2 dk 00 sn');
  });

  /** ⚠️ Ama BAKIMDA ayrışıyorlar — ve geri sayım donmuş saatten çizilmeli. */
  it('bakımda geri sayım DONMUŞ saatten çizilir', () => {
    const pausedAt = BROWSER_NOW - 10 * 60_000;
    noteServerTime(iso(BROWSER_NOW), iso(pausedAt));
    const varis = iso(pausedAt + 120_000);           // duraklama anından 2 dk sonrası

    expect(remaining(varis)).toBe('2 dk 00 sn');     // ⭐ donmuş: iş ilerlemiyor
    expect(remaining(varis, serverNow())).toBeNull(); // ⛔ gerçek saat: "bitmiş" sanır
  });

  it('bitmiş süre null döner', () => {
    expect(remaining(iso(gameNow() - 1_000))).toBeNull();
  });

  it('boş/geçersiz girdi null döner', () => {
    expect(remaining(null)).toBeNull();
    expect(remaining(undefined)).toBeNull();
    expect(remaining('çöp')).toBeNull();
  });

  it('remainingClock aynı çıpayı kullanır', () => {
    noteServerTime(iso(BROWSER_NOW), iso(BROWSER_NOW - 196_563));
    expect(remainingClock(iso(gameNow() + 271_000))).toBe('04:31');
  });

  /**
   * ⚠️ `remainingLong` varsayılanı GERÇEK saat — çünkü tek çağıranı tatil paneli ve oradaki
   * damga gün ölçeğinde. Çıpa isteyen açıkça geçmeli. Bu testin işi o asimetriyi **bilinçli**
   * kılmak: değişirse burada görünür.
   */
  it('remainingLong varsayılanı GERÇEK saat (bilinçli asimetri)', () => {
    noteServerTime(iso(BROWSER_NOW), iso(BROWSER_NOW - 196_563));
    const gunler = iso(serverNow() + 3 * 86_400_000);
    expect(remainingLong(gunler)).toBe('3 gün');
  });
});

/**
 * ⭐ GEÇMİŞE BAKAN DAMGA — sohbetlerin zaman gösterimi (kullanıcı, 2026-08-11).
 *
 * ⚠️ Eskiden yalnız `SS:dd` yazıyordu: üç gün önce 13:06'da gelen mesajla bugün 13:06'da
 * gelen mesaj ekranda AYNI görünüyordu (kullanıcının bildirdiği hata).
 */
describe('timeAgo — geçmişe bakan süre', () => {
  const gecmis = (ms: number): string => iso(BROWSER_NOW - ms);

  it('eşikleri sırayla geçer: saniye → dakika → saat → gün → ay → yıl', () => {
    expect(timeAgo(gecmis(3_000))).toBe('3 saniye önce');
    expect(timeAgo(gecmis(59_000))).toBe('59 saniye önce');
    expect(timeAgo(gecmis(60_000))).toBe('1 dakika önce');
    expect(timeAgo(gecmis(45 * 60_000))).toBe('45 dakika önce');
    expect(timeAgo(gecmis(3_600_000))).toBe('1 saat önce');
    expect(timeAgo(gecmis(23 * 3_600_000))).toBe('23 saat önce');
    expect(timeAgo(gecmis(86_400_000))).toBe('1 gün önce');
    expect(timeAgo(gecmis(29 * 86_400_000))).toBe('29 gün önce');
    expect(timeAgo(gecmis(30 * 86_400_000))).toBe('1 ay önce');
    expect(timeAgo(gecmis(200 * 86_400_000))).toBe('6 ay önce');
    expect(timeAgo(gecmis(400 * 86_400_000))).toBe('1 yıl önce');
  });

  /**
   * ⚠️⚠️ ÇIPA `serverNow()` — ve bu, `remaining` ailesinin TERSİ. Sebebi
   * `chat_messages.created_at`in `NON_TIMELINE_COLUMNS` içinde olması: sohbet damgası bir oyun
   * olayı değil, geçmiş kaydıdır ve bakım kaydırmasından etkilenmez. `gameNow()` kullansaydık
   * bakım boyunca bütün mesajların yaşı donardı.
   */
  it('⭐ bakımda bile YAŞLANMAYA DEVAM eder (gameNow değil serverNow)', () => {
    const pausedAt = BROWSER_NOW - 10 * 60_000;
    noteServerTime(iso(BROWSER_NOW), iso(pausedAt));
    expect(gameNow()).not.toBe(serverNow());          // saatler gerçekten ayrışmış

    const mesaj = iso(BROWSER_NOW - 5 * 60_000);
    expect(timeAgo(mesaj)).toBe('5 dakika önce');
    // ⛔ Oyun saatiyle bakılsaydı mesaj «gelecekte» kalır ve «1 saniye önce» donardı:
    expect(timeAgo(mesaj, gameNow())).toBe('1 saniye önce');
  });

  /**
   * ⚠️ GELECEK damga «1 saniye önce» sayılır, negatif YAZILMAZ: sunucu ile tarayıcı saati
   * arasındaki küçük kayma yüzünden yeni gönderilmiş bir mesaj birkaç saniye ileride
   * görünebiliyor ve *"-2 saniye önce"* kabul edilemez.
   */
  it('⚠️ gelecek damga ve 0 saniye «1 saniye önce» olur, negatife düşmez', () => {
    expect(timeAgo(iso(BROWSER_NOW))).toBe('1 saniye önce');
    expect(timeAgo(iso(BROWSER_NOW + 4_000))).toBe('1 saniye önce');
  });

  it('boş/geçersiz girdi tire döner (geri sayımlardaki `null` DEĞİL)', () => {
    expect(timeAgo(null)).toBe('—');
    expect(timeAgo(undefined)).toBe('—');
    expect(timeAgo('çöp')).toBe('—');
  });
});

describe('biçimlendirme', () => {
  it('formatDuration saniyeyi HER ZAMAN yazar (kullanıcı kararı)', () => {
    expect(formatDuration(7_457)).toBe('2 sa 04 dk 17 sn');
    expect(formatDuration(192)).toBe('3 dk 12 sn');
    expect(formatDuration(9)).toBe('9 sn');
    expect(formatDuration(-5)).toBe('0 sn');
  });

  it('formatLongDuration bir günden kısasını formatDuration\'a devreder', () => {
    expect(formatLongDuration(86_399)).toBe('23 sa 59 dk 59 sn');
    expect(formatLongDuration(86_400)).toBe('1 gün');
    expect(formatLongDuration(30 * 86_400 - 60)).toBe('29 gün 23 sa');
  });

  it('formatClock dar yerler için saat biçimi verir', () => {
    expect(formatClock(271)).toBe('04:31');
    expect(formatClock(7_467)).toBe('2:04:27');
  });
});

File: apps/mobile/test/core/clock_test.dart
/// ⭐⭐ İSTEMCİ SAATİ — durum davranışı + DİLLER ARASI EŞİTLİK KAPISI.
///
/// İki bölüm, iki ayrı iş:
///
///   1. **Durum** (`saat çıpaları`, `remaining`) — web'deki `apps/web/test/clock.test.ts`in
///      aynadaki karşılığı. Sapma ölçümü, bakım donması, çıpa seçimi.
///   2. ⭐⭐ **Ortak vektörler** — `packages/contracts/fixtures/clock-vectors.json` dosyasını
///      TS testiyle **aynı dosyadan** okuyup aynı sonuçları bekliyor. Kullanıcının «tam
///      eşitlik» kararı böylece umut değil kapı oluyor.
///
/// ⚠️ Vektör dosyası KOPYALANMIYOR, göreli yoldan okunuyor (`MOBIL_MIMARI.md` §5.1): kopya,
/// öldürmeye çalıştığımız bayatlamanın ta kendisi olurdu.
library;

import 'dart:convert';
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';
import 'package:mobilwar/core/clock.dart';

/// ⚠️ `flutter test` çalışma dizini paket kökü (`apps/mobile`) — yol oradan kuruluyor.
final _vectors =
    jsonDecode(
          File(
            '../../packages/contracts/fixtures/clock-vectors.json',
          ).readAsStringSync(),
        )
        as Map<String, dynamic>;

List<Map<String, dynamic>> _grup(String ad) =>
    (_vectors[ad] as List<dynamic>).cast<Map<String, dynamic>>();

/// Sabit bir "cihaz şimdisi" — sapma ölçümü ancak böyle deterministik olur.
const int _cihazSimdi = 1786060800000; // 2026-08-07T12:00:00.000Z

String _iso(int ms) =>
    DateTime.fromMillisecondsSinceEpoch(ms, isUtc: true).toIso8601String();

/// Saati sabit cihaz anına çıpalanmış, dünyası ÇALIŞAN bir saat kurar.
MwClock _saat({int? simdi}) {
  var an = simdi ?? _cihazSimdi;
  final c = MwClock(deviceNow: () => an);
  c.noteServerTime(_iso(an), _iso(an));
  return c;
}

void main() {
  // ── Bölüm 1: durum davranışı ────────────────────────────────────────────────
  group('saat çıpaları', () {
    test('⚠️ sabit doğru okundu (vektör dosyasıyla aynı sayı)', () {
      expect(kPauseThresholdMs, _vectors['pauseThresholdMs']);
    });

    test('cihaz saati ileriyse sapma ölçülüp geri alınır', () {
      final c = MwClock(deviceNow: () => _cihazSimdi);
      // Sunucu 30 sn GERİDE → cihaz 30 sn ileri demektir.
      c.noteServerTime(_iso(_cihazSimdi - 30000));
      expect(c.serverNow(), _cihazSimdi - 30000);
    });

    test('cihaz saati geriyse de düzeltilir', () {
      final c = MwClock(deviceNow: () => _cihazSimdi);
      c.noteServerTime(_iso(_cihazSimdi + 45000));
      expect(c.serverNow(), _cihazSimdi + 45000);
    });

    test('⭐⭐ dünya çalışıyorken gameNow == serverNow (tek zaman çizgisi)', () {
      final c = _saat();
      expect(c.gameNow(), c.serverNow());
    });

    test('⭐ bakımda gameNow DONAR, serverNow AKMAYA DEVAM eder', () {
      var an = _cihazSimdi;
      final c = MwClock(deviceNow: () => an);
      final pausedAt = _cihazSimdi - 10 * 60000;
      c.noteServerTime(_iso(_cihazSimdi), _iso(pausedAt));
      expect(c.gameNow(), pausedAt);

      an = _cihazSimdi + 30000; // gerçek zaman aksın
      expect(c.serverNow(), _cihazSimdi + 30000);
      expect(c.gameNow(), pausedAt, reason: 'oyun saati kıpırdamamalı');
    });

    test('⚠️ gameNow göndermeyen yanıt bakım çıpasını BOZMAZ', () {
      // Uçların çoğu `serverNow` gönderiyor ama `gameNow` göndermiyor; her biri perdeyi
      // sessizce kaldırsaydı bakım sırasında geri sayımlar rastgele akıp dururdu.
      final c = MwClock(deviceNow: () => _cihazSimdi);
      final pausedAt = _cihazSimdi - 60000;
      c.noteServerTime(_iso(_cihazSimdi), _iso(pausedAt));
      c.noteServerTime(_iso(_cihazSimdi)); // gameNow yok
      expect(c.gameNow(), pausedAt);
    });

    test('⚠️ bozuk damga çıpayı bozmaz', () {
      final c = _saat();
      c.noteServerTime('bu bir tarih değil', 'bu da değil');
      expect(c.gameNow(), c.serverNow());
    });

    test('noteMaintenance anında dondurur ve çözer', () {
      final c = _saat();
      final pausedAt = _cihazSimdi - 5000;
      c.noteMaintenance(true, _iso(pausedAt));
      expect(c.gameNow(), pausedAt);

      c.noteMaintenance(false);
      expect(c.gameNow(), c.serverNow());
    });

    test('damgasız bakım bildirimi yine de dondurur', () {
      var an = _cihazSimdi;
      final c = MwClock(deviceNow: () => an);
      c.noteServerTime(_iso(an), _iso(an));
      c.noteMaintenance(true);
      final donmus = c.gameNow();

      an = _cihazSimdi + 10000;
      expect(
        c.gameNow(),
        donmus,
        reason: 'yanlış an, ama akmaya devam etmekten iyi',
      );
    });
  });

  group('remaining — çıpa seçimi', () {
    test('⭐⭐ 2026-08-02 hatasının bekçisi: iki çıpa da aynı sonucu verir', () {
      // Casusluk 120 sn sürüyor. Oyun saati gerçek saatin gerisinde kalsaydı varış anı gerçek
      // saatle bakıldığında HEP geçmişte kalır ve geri sayım yerine sürekli «varıyor» yazardı.
      final c = _saat();
      final varis = _iso(c.gameNow() + 120000);

      expect(c.remaining(varis), '2 dk 00 sn');
      expect(c.remaining(varis, now: c.serverNow()), '2 dk 00 sn');
    });

    test('⚠️ BAKIMDA geri sayım DONMUŞ saatten çizilir', () {
      final c = MwClock(deviceNow: () => _cihazSimdi);
      final pausedAt = _cihazSimdi - 10 * 60000;
      c.noteServerTime(_iso(_cihazSimdi), _iso(pausedAt));
      final varis = _iso(pausedAt + 120000); // duraklama anından 2 dk sonrası

      expect(
        c.remaining(varis),
        '2 dk 00 sn',
        reason: '⭐ donmuş: iş ilerlemiyor',
      );
      expect(
        c.remaining(varis, now: c.serverNow()),
        isNull,
        reason:
            '⛔ gerçek saatle bakılsaydı "bitmiş" sanırdı — tam olarak o hata',
      );
    });

    test('bitmiş süre ve geçersiz girdi null döner', () {
      final c = _saat();
      expect(c.remaining(_iso(c.gameNow() - 1000)), isNull);
      expect(c.remaining(null), isNull);
      expect(c.remaining('çöp'), isNull);
    });

    test('remainingClock aynı çıpayı kullanır', () {
      final c = _saat();
      expect(c.remainingClock(_iso(c.gameNow() + 271000)), '04:31');
    });

    test('⚠️ remainingLong varsayılanı GERÇEK saat (bilinçli asimetri)', () {
      // Tek çağıranı tatil paneli ve oradaki damga gün ölçeğinde. Ailenin geri kalanından
      // ayrışması kasıtlı; değişirse burada görünür.
      final c = MwClock(deviceNow: () => _cihazSimdi);
      c.noteServerTime(_iso(_cihazSimdi), _iso(_cihazSimdi - 196563));
      expect(c.remainingLong(_iso(c.serverNow() + 3 * 86400000)), '3 gün');
    });

    test('⭐ timeAgo çıpası serverNow — bakımda bile YAŞLANMAYA devam eder', () {
      final c = MwClock(deviceNow: () => _cihazSimdi);
      final pausedAt = _cihazSimdi - 10 * 60000;
      c.noteServerTime(_iso(_cihazSimdi), _iso(pausedAt));
      expect(c.gameNow(), isNot(c.serverNow())); // saatler gerçekten ayrışmış

      final mesaj = _iso(_cihazSimdi - 5 * 60000);
      expect(c.timeAgo(mesaj), '5 dakika önce');
      // ⛔ Oyun saatiyle bakılsaydı mesaj «gelecekte» kalır ve donardı:
      expect(c.timeAgo(mesaj, now: c.gameNow()), '1 saniye önce');
    });
  });

  // ── Bölüm 2: ⭐⭐ diller arası eşitlik ───────────────────────────────────────
  group('⭐⭐ ortak vektörler — web ile AYNI dosya, AYNI sonuç', () {
    test('⚠️ vektör dosyası gerçekten okundu (sessiz boş küme değil)', () {
      // Dosya bulunamasa `readAsStringSync` zaten patlar; asıl risk yanlış anahtar okuyup
      // boş liste üzerinde dönmek — o durumda döngülü testler HİÇBİR ŞEY ölçmeden yeşil yanar.
      expect(_grup('durations').length, greaterThan(10));
      expect(_grup('timeAgo').length, greaterThan(10));
      expect(_grup('pause').length, greaterThan(3));
    });

    for (final v in _grup('durations')) {
      test('${v['seconds']} sn → ${v['duration']}', () {
        final s = v['seconds'] as num;
        expect(formatDuration(s), v['duration']);
        expect(formatLongDuration(s), v['long']);
        expect(formatClock(s), v['clock']);
      });
    }

    for (final v in _grup('timeAgo')) {
      test('${v['agoSeconds']} sn önce → ${v['text']}', () {
        final ago = (v['agoSeconds'] as num).toInt();
        final c = _saat();
        expect(c.timeAgo(_iso(_cihazSimdi - ago * 1000)), v['text']);
      });
    }

    for (final v in _grup('pause')) {
      test('sapma ${v['skewMs']} ms → duraklamış: ${v['paused']}', () {
        final skew = (v['skewMs'] as num).toInt();
        final c = _saat(); // önce dünyayı çalışır duruma çek
        c.noteServerTime(_iso(_cihazSimdi), _iso(_cihazSimdi - skew));

        if (v['paused'] as bool) {
          expect(c.gameNow(), _cihazSimdi - skew);
        } else {
          expect(c.gameNow(), c.serverNow());
        }
      });
    }
  });
}
